### PR TITLE
SelectAcceleratorCode: strip pointer casts when examining a call

### PR DESCRIPTION
**Pull request by ssahasra**
At a call-site, the callee may not directly be a pointer to the
function; it can also be a cast expression even that points to the
actual function. This can happen even without an indirect call. The
traversal of the call-graph needs to be aware of such pointer casts.

### CODE REVIEW

**Comment by ssahasra** [comment]
@AlexVlx could you please take a look at this merge request? This is a simple way to handle function calls where the argument types don't match due to aliasing of global and generic pointers in HSA. The heavy gun would be run instcombine, but this is too early in the flow for that.

FYI: @b-sumner

**Comment by AlexVlx** [comment]
@ssahasra, sorry I missed this:
a) I retain the objections I expressed elsewhere about allowing bitcasted calls in general, especially since what you're addressing is PRECISELY what HLLs define as UB - I'd have rather preferred if ROCDL was truly generic and bit the bullet of doing address space casts in the definition of a function, leaving the interface untainted by explicit address spaces, however that's a different kettle of fish; however, nobody else seems particularly bothered, so this is fine on that account;
b) The convergent / OptNone noise should go elsewhere, be it a different pass or a different function with some context as to why this is done, so that future generations who were not there when "Matt said so" can at least have a hope of understanding the why; also, hopefully this pass is going away sooner rather than later, so you'll have to find a new house anyway.
So modulo (b) I'm fine with this.

**Comment by ssahasra** [comment]
Oops. The convergent "noise" was not meant for your eyes! Moved it elsewhere.

Regarding (a), I have found no evidence that the bitcast invokes UB. On the contrary, it seems to be a well-supported pattern that is handled correctly and explicitly by InstCombine. Even if it really were UB, allowing the bitcast in the select... pass does not introduce any new bug in the program. It just means that an occasional function that should have been replaced with undef, will instead make it to the list of device functions. Both situations are covered by the UB, if that's what it is.

LGTY?

**Comment by ssahasra** [comment]
Bump! I would like to close this ASAP, so that I can move on to the next change to be reviewed!

**Comment by AlexVlx** [comment]
You still need a second review IIRC. I already said that without the noise it's fine as far as I'm concerned, since I do not see the discussion about UB as going anywhere and we'll just have to agree to disagree. So can you ask @scchan to review so that you can move on to the next change to be reviewed? Thanks.

**Review by AlexVlx — APPROVED** [review]

**Comment by ssahasra** [comment]
ping: @scchan

**Review by scchan — APPROVED** [review]
LGTM

**Pull request merged**
